### PR TITLE
feat(router/atc): support field `http.path.segments.len`

### PR DESCRIPTION
**Pull request by chronolaw**
### Summary

[KAG-3604]

### Checklist

- [x] The Pull Request has tests
- [x] A changelog file has been created under `changelog/unreleased/kong` or `skip-changelog` label added on PR if changelog is unnecessary. [README.md](https://github.com/Kong/gateway-changelog/README.md)
- [ ] There is a user-facing docs PR against https://github.com/Kong/docs.konghq.com - PUT DOCS PR HERE

### Issue reference

Fix #_[issue number]_

[KAG-3604]: https://konghq.atlassian.net/browse/KAG-3604?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: kong/router/fields.lua [hunk 1]
```diff
@@ -298,6 +299,13 @@ if is_http then
         -- http.path.segments.0 => params.segments[1 + 0] => a
         -- http.path.segments.1_2 => b/c
 
+        if range == "len" then
+          value = #segments
+          segments[range] = value
+
+          return value
+        end
+
```
Comment (dndx) [inline kong/router/fields.lua:308]: We can put `http.path.segments.len` directly into `FIELDS_FUNCS`, this should override the `http.path.segments.*` function.

Comment (chronolaw) [reply kong/router/fields.lua:308]: I prefer to implement here, because we will split the uri, or else we have to repeat the split code.

Comment (dndx) [reply kong/router/fields.lua]: Can you do a `get_segments(param)` instead? We can simply do a refactor to load and cache the segments. Logically it is clearer than calling `len` the range.

Comment (chronolaw) [reply kong/router/fields.lua]: done

File: kong/db/schema/entities/routes.lua [hunk 2]
```diff
@@ -27,10 +27,11 @@ do
...
     for _, f in ipairs(fields) do
       if f:find(HTTP_PATH_SEGMENTS_PREFIX, 1, true) then
-        local m = re_match(f:sub(#HTTP_PATH_SEGMENTS_PREFIX + 1),
-                           HTTP_PATH_SEGMENTS_SUFFIX_REG, "jo")
+        local suffix = f:sub(#HTTP_PATH_SEGMENTS_PREFIX + 1)
+        local m = re_match(suffix, HTTP_PATH_SEGMENTS_SUFFIX_REG, "jo")
 
-        if not m or (m[2] and tonumber(m[1]) >= tonumber(m[3])) then
+        if (suffix ~= "len") and
+           (not m or (m[2] and tonumber(m[1]) >= tonumber(m[3]))) then
```
Comment (chobits) [inline kong/db/schema/entities/routes.lua:34]: hi @chronolaw Im curious what coding style should do when we need to fold the cold line

```
if ... and/or
   ... and/or
   ... and/or ... then
...
end

or

if ... and/or
   ... and/or
   ... and/or ...
then
...
end

```

Comment (chronolaw) [reply kong/db/schema/entities/routes.lua:34]: It seems that now we have no precisely rule for it, each style is ok.

File: kong/router/fields.lua [hunk 3]
```diff
@@ -213,6 +214,24 @@ if is_http then
...
+      HTTP_SEGMENTS_REG_CTX.pos = 2 -- reset ctx, skip first '/'
+      params.segments = re_split(params.uri, "/", "jo", HTTP_SEGMENTS_REG_CTX)
+    end
+
+    return params.segments
+  end
+
+
+  FIELDS_FUNCS["http.path.segments.len"] =
+  function(params)
```
Comment (chobits) [inline kong/router/fields.lua:228]: seems that we always use the following style in everywhere of kong

```
<function variable / ...> = function           -- function key word is in the assignment line
...
end
```

File: kong/db/schema/entities/routes.lua (hunk 2 above)
Comment (chobits) [reply kong/db/schema/entities/routes.lua:34]: KK i see

**Review by chobits — APPROVED** [review]

File: kong/router/fields.lua (hunk 3 above)
Comment (chronolaw) [reply kong/router/fields.lua:228]: Here we add a function to the hash table, it is not a normally function definition.

And sometimes `local f = function()... end` may show the intention that the `f` is overwrited.

**Review by ADD-SP — APPROVED** [review]

File: kong/router/fields.lua [hunk 4]
```diff
@@ -213,6 +214,24 @@ if is_http then
   local HTTP_SEGMENTS_OFFSET = 1
```
Comment (dndx) [inline kong/router/fields.lua:216]: Move `HTTP_SEGMENTS_REG_CTX` to here:

```
do
  local HTTP_SEGMENTS_REG_CTX = xxx

  local function get_http_segments(params)
  ...
  end
end
```

Comment (chronolaw) [reply kong/router/fields.lua:215]: done.

**Review by dndx — APPROVED** [review]

**Pull request merged**